src/control/server/iosrv.go

```diff
@@ -513,34 +513,56 @@ func (srv *iosrv) writeSuper() error {
 	return writeIosrvSuper(iosrvSuperPath(srv.config.Servers[srv.index].ScmMount), srv.super)
 }
 
-// mgmtJoin sends the Join request to MS, retrying indefinitely until MS
-// responses.
-func mgmtJoin(ap string, tc *security.TransportConfig, req *mgmtpb.JoinReq) (*mgmtpb.JoinResp, error) {
-	log.Debugf("join(%s, %+v)", ap, *req)
+// mgmtJoin sends Join request(s) to MS.
+func mgmtJoin(ctx context.Context, ap string, tc *security.TransportConfig, req *mgmtpb.JoinReq) (*mgmtpb.JoinResp, error) {
+	prefix := fmt.Sprintf("join(%s, %+v)", ap, *req)
+	log.Debugf(prefix + " begin")
+	defer log.Debugf(prefix + " end")
+
+	const delay = 3 * time.Second
+
 	var opts []grpc.DialOption
 	authDialOption, err := security.DialOptionForTransportConfig(tc)
 	if err != nil {
 		return nil, errors.Wrap(err, "Failed to determine dial option from TransportConfig")
 	}
 
 	// Setup Dial Options that will always be included.
-	opts = append(opts, grpc.WithBlock(), grpc.WithBackoffMaxDelay(5*time.Second),
+	opts = append(opts, grpc.WithBlock(), grpc.WithBackoffMaxDelay(delay),
 		grpc.WithDefaultCallOptions(grpc.FailFast(false)), authDialOption)
-	conn, err := grpc.Dial(ap, opts...)
+	conn, err := grpc.DialContext(ctx, ap, opts...)
 	if err != nil {
 		return nil, errors.Wrapf(err, "dial %s", ap)
 	}
 	defer conn.Close()
 
 	client := mgmtpb.NewMgmtSvcClient(conn)
 
-	resp, err := client.Join(context.Background(), req)
-	if err != nil {
-		return nil, errors.Wrapf(err, "join %s %v", ap, *req)
-	}
-	if resp.Status != 0 {
-		return nil, errors.Errorf("join %s %v: %d\n", ap, *req, resp.Status)
-	}
+	for {
+		select {
+		case <-ctx.Done():
+			return nil, errors.Wrap(ctx.Err(), prefix)
+		default:
+		}
 
-	return resp, nil
+		resp, err := client.Join(ctx, req)
+		if err != nil {
+			log.Debugf(prefix+": %v", err)
+		} else {
+			// TODO: Stop retrying upon certain errors (e.g., "not
+			// MS", "rank unavailable", and "excluded").
+			if resp.Status != 0 {
+				log.Debugf(prefix+": %d", resp.Status)
+			} else {
+				return resp, nil
+			}
+		}
+
+		// Delay next retry.
+		delayCtx, cancelDelayCtx := context.WithTimeout(ctx, delay)
```

Comment:
FWIW, I think you could also just block on <-time.After(delay) here and avoid the delayCtx stuff. https://golang.org/pkg/time/#After